### ISSUE TITLE
word break cause by color tag

### ISSUE THREAD

**Issue by loohoo** [libgdx/libgdx#4672]
#### Issue details
the second draw of font should not break the word "international".  It just one word with different color.

#### Reproduction steps/code
```
	public void render(float delta) {
		Gdx.gl.glClearColor(0.6f, 0.6f, 0.6f, 1.0f);
		Gdx.gl.glClear(GL20.GL_COLOR_BUFFER_BIT);
		
		batch.setProjectionMatrix(camera.combined);

		shape.begin(ShapeRenderer.ShapeType.Line);
		shape.setColor(Color.RED);
		shape.rect(0, VIRTUAL_HEIGHT - 600, 300, 600);
		shape.end();

		batch.begin();
		font.getData().setScale(0.5f);
		font.draw(batch, "[BLUE]Is this a bug? international string123456789abcdefghiklmn",
				0.0f, VIRTUAL_HEIGHT,
				300.f, Align.left, true);
              // [output](http://pasteboard.co/OPDave7xa.png)
		font.draw(batch, "[BLUE]Is this a bug? inter[RED]national string123456789abcdefghiklmn",
				0.0f, VIRTUAL_HEIGHT - 200,
				300.f, Align.left, true);
               // [output](http://pasteboard.co/OPuY5FCP6.png)
		batch.end();

	}


```

#### Version of LibGDX and/or relevant dependencies
libgdx 1.9.5

**Comment by TCreutzenberg** [comment]
@NathanSweet @tommyettinger If you guys are not on this already I would have a look at it tomorrow. I wanted to fix this tomorrow actually anyways and didn't even know there was an issue about this already 😄

**Comment by NathanSweet** [comment]
I'm not on it. Best of luck! thar be dragonz

**Comment by TCreutzenberg** [comment]
Hey guys, I need some opinions to actually be able to fix this bug.

GlyphLayout#setText() at the moment does many things in one single loop:
- Adding glyphs
- Checking linebreaks and color-markups
- Truncating + (non-color-markup) line wrapping
- Handling all the special cases for the above

It does this by reusing all the same variables all the time. This makes it incredibly hard to follow and debug. And it also connects all the above tasks directly to another via the shared variables, making it even worse to debug.

If I add correct line wrapping for color-markup (which would be possible I believe) it will be very very very hard for other people to ever understand what's going in the code. You already need to know a lot about fonts and you have to understand all the GlyphLayout/GlyphRun/Truncate/Wrap/Color-markup logic and how they interact (!!) already.

To actually make this thing "maintainable" it would be necessary to break it up into several steps. It might loose a little performance (even if I doubt it would be really relevant at all / possible to prove it in a alive test) but it would be possible handle less "tasks" at a time and make it easier to understand what's going on. 

Propositions would be:

1) Do everything in "the one loop"
Pro: Maybe bit faster (if all all)
Con: A hell to maintain, debug and to understand

2) Parse all GlyphRuns first (+ truncate if applicable), then do wrapping in an extra loop
Pro: Easier logic, separation of concerns in the code, making it easier to add new features or to debug
Con: Might be a bit slower (if all all)

There are several options to parse the GlyphRuns
a) One GlyphRun per word/color-markup
Pro: Easy line wrapping and to understand/maintain
Con: Needs more GlyphRuns (but they would be shorter!)
b) GlyphRuns can span several words
Pro: Less GlyphRuns (but they would be longer!)
Con: More work/effort to line-wrap than a)

What are your thoughts on this? I don't want to do a days work just to find out you guys don't like the idea.

My personal preference would be 2.a) But I'm also happy to implement any other proposition as long as it's doable.

@NathanSweet @tommyettinger What are your preferences?

**Comment by NathanSweet** [comment]
It's hard to judge how much of an impact such refactoring would have on performance. I can say when starting Spine, GlyphLayout#setText is called 4723 times. After start up it's called around 2-20 times per frame. I'm not super excited about making it slower so the code can be easier to follow. I'm most interested in the code doing the right things, and quickly.

There are many edge cases in text layout. The current GlyphLayout code has had issues ironed out for a long time and is pretty solid. I'm not sure it's worth starting over and going through the inevitable new problems, performance implications aside.

As for code readability, I'm sure it's possible to make it nicer as you say, maybe by doing it in multiple passes. However, I'm also pretty sure it would still be quite a monster in the end due to the number of edges cases and the complexity needed to handle them all.

The code used to do everything in one pass, eg keeping a running calculation of the widths, adjusting as needed. Relatively recently this was changed to go over all the runs at the end. Maybe it could be adjusted to create the runs, then do the wrapping. It's not clear how much better this will be though, as much of the same work needs to be done.

I haven't looked at what would be needed to prevent wrapping at color tags. I have a feeling it shouldn't be too hard using the existing code, enow that we have kerning between runs (that didn't used to be the case). I'm inclined to explore that before risking instability with a larger refactoring.

**Comment by tommyettinger** [comment]
Yeah, it might be worth it to look at how kerning handles the interaction between runs, since that should be similar to how runs could be wrapped as one unit (or split on a word boundary rather than a run boundary). I'm also definitely in favor of fast code over readable code in this case, because text layout is rarely especially readable even in the best case, but it needs to be fast for text-heavy applications. We also hopefully won't need to read this code especially often as any remaining bugs are ironed out (wishful thinking, ha).

**Comment by NathanSweet** [comment]
If we didn't need this damn feature to not wrap color runs, we could call it perfect as is! :D

Giving it some thought, it'd go like this:

We place a run that fits on a line, we move to the next run, it doesn't fit! We need to decide where to break the run to wrap it. Currently if it doesn't have a break char, we move the whole run to the next line.

Instead of doing that, we need to look at the runs before it (potentially all of them on the same line) for a break char. If we find one, that's the run we need to break. Afterward, we need to reposition any runs after it (just position, don't need `getGlyphs` again) until we get back to the original run that didn't fit, then we continue placing runs on the next line.

That sounds a little tricky when written out, but may not be too bad in code (and by that I mean just as bad as the other code!).

**Comment by TCreutzenberg** [comment]
@NathanSweet Yes, what you wrote up is exactly what I was going to try. Maybe it's even possible to store the last potential linebreak-position/index while parsing the runs so you mustn't look where one is later. And `fontData.getWrapIndex(run.glyphs, i)` goes back the whole run/s anyways.

**Comment by NathanSweet** [comment]
Cool!

I'm not sure about keeping track of the last possible break position. `getWrapIndex` can be overridden to customize the logic. CJK each use different rules and need more logic than `isWhitespace` and `isBreakChar`. For example: http://n4te.com/x/2057-6deb.txt

**Comment by TCreutzenberg** [comment]
Thanks for the info! I already started it wrong then... It's actually harder than I thought...

**Comment by NathanSweet** [comment]
Not doing it is an option! ;)

**Comment by TCreutzenberg** [comment]
Nah, I spent so much time already, I'm not going to resign.

I already abandoned the option we talked about above as it turned out to be really crazy because recursive linebreaks are needed where you have to keep track of all runs which are already dangling behind the one to wrap, etc. Not really a sane option, I tried it so I know. Or I just didn't get it working correctly, also possible.

Whenever something is overly complicated I often remember that in these cases normally the underlying data structure or architecture is not suited to the problem to be solved. After looking at the whole thing the problem with the data structure here is that wrapping is mixed with color changes. I.e. lines are broken because of color changes. That's the reason why wrapping is so complicated.
There is a really cool way to solve this problem I'm working on right now which would require the GlyphRuns to change the color-change handling a little bit, the rest of the code would even be simplified a lot and still using the all the methods which are already in use. It'll basically will be only one GlyphRun per line.
If GlyphRun color handling is not to be touched this can also changed to create GlyphRuns as they are now with a little overhead.

When I'm finished, please test with your project if you notice any difference. Even if you guys do not like it I will still use it for my own project, as I'll be using color-markup in many texts I guess.

**Comment by NathanSweet** [comment]
Closed by #6587
